### PR TITLE
core: Option to require msgrv2 even without encryption

### PR DESCRIPTION
**Pull request by travisn**
**Description of your changes:**
Enabling msgr v2 and disabling msgr v1 currently requires enabling either encryption on the wire or compression on the wire. As more clients are running on the latest kernel, allow the clients to run on v2 even when encryption and compression are not enabled. Clusters that are fully running on v2 will more easily be able to change configuration between enabling or disabling msgr v2 features.

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by sp98 — APPROVED** [review]

File: deploy/examples/cluster.yaml [hunk 1]
```diff
@@ -88,6 +88,10 @@ spec:
       # Requires Ceph Quincy (v17) or newer. Also see the kernel requirements above for encryption.
       compression:
         enabled: false
+      # Whether to require communication over msgr v2. If true, the msgr v1 port will be disabled
```
Comment (Madhu-1) [inline deploy/examples/cluster.yaml:91]: Not possible to configure encryption support in helm values.yaml?

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 2]
```diff
@@ -2170,6 +2170,12 @@ type ConnectionsSpec struct {
 	// +nullable
 	// +optional
 	Compression *CompressionSpec `json:"compression,omitempty"`
+
+	// Whether to require msgr2 even if compression or encryption are not enabled.
+	// If true, the msgr1 port will be disabled.
+	// Requires a kernel that supports msgr v2 (mainline kernel 5.11 or CentOS 8.4).
+	// +optional
+	RequireMsgrV2 bool `json:"requireMsgrV2,omitempty"`
```
Comment (Madhu-1) [inline pkg/apis/ceph.rook.io/v1/types.go:2178]: should we call this msgrv2 or just msgr2 as per ceph https://docs.ceph.com/en/quincy/dev/msgr2/?

Comment (sp98) [reply pkg/apis/ceph.rook.io/v1/types.go:2178]: +1 on `msgr2`. Rook already calls it `msgr2`

File: deploy/examples/cluster.yaml (hunk 1 above)
Comment (travisn) [reply deploy/examples/cluster.yaml:91]: Yes, anything in the cluster spec can be added in the `cephClusterSpec` of the cluster chart. I'll add it there just to be clear, thanks.

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 2 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:2178]: Agreed on the consistency, will update it

**Comment by travisn** [comment]
@Madhu-1 @malayparida2000 A new commit is pushed for the behavior discussed... Mons that are configured on msgr1 will be using the v1 port in the csi driver since only one port is allowed in the configmap. To allow mons to be started on v2 while keeping quorum with the v1 mons, the v1 mons must continue running on the v1 port.  So even if msgr2 is required, allow existing v1 mons to continue running on that port.

For a cluster on msgr1 to enable msgr2 with encryption or compression, it would require these steps:
1. Set `network.connections.requireMsgr2: true`. After this, new mons will only be started on the v2 port, but existing mons will remain on v1
2. Failover the mons 1 by 1 until they are all replaced (scale down a mon and wait 10 minutes for it to be failed over)

**Comment by malayparida2000** [comment]
> @Madhu-1 @malayparida2000 A new commit is pushed for the behavior discussed... Mons that are configured on msgr1 will be using the v1 port in the csi driver since only one port is allowed in the configmap. To allow mons to be started on v2 while keeping quorum with the v1 mons, the v1 mons must continue running on the v1 port. So even if msgr2 is required, allow existing v1 mons to continue running on that port.
> 
> For a cluster on msgr1 to enable msgr2 with encryption or compression, it would require these steps:
> 
> 1. Set `network.connections.requireMsgr2: true`. After this, new mons will only be started on the v2 port, but existing mons will remain on v1
> 2. Failover the mons 1 by 1 until they are all replaced (scale down a mon and wait 10 minutes for it to be failed over)

Sounds interesting, So does this mean for some time we may see twice the number of mons compared to what we see today?

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
> @Madhu-1 @malayparida2000 A new commit is pushed for the behavior discussed... Mons that are configured on msgr1 will be using the v1 port in the csi driver since only one port is allowed in the configmap. To allow mons to be started on v2 while keeping quorum with the v1 mons, the v1 mons must continue running on the v1 port. So even if msgr2 is required, allow existing v1 mons to continue running on that port.
> 
> For a cluster on msgr1 to enable msgr2 with encryption or compression, it would require these steps:
> 
> 1. Set `network.connections.requireMsgr2: true`. After this, new mons will only be started on the v2 port, but existing mons will remain on v1
> 2. Failover the mons 1 by 1 until they are all replaced (scale down a mon and wait 10 minutes for it to be failed over)

@travisn sounds good are we documenting this one? If yes we need to document the behavior of already mounted csi volumes also.

**Comment by travisn** [comment]
> Sounds interesting, So does this mean for some time we may see twice the number of mons compared to what we see today?

To follow up from discussion, the most mons ever seen during a failover is one extra, then the failed mon is removed by rook and returns to three mons in quorum.

**Comment by travisn** [comment]
> @travisn sounds good are we documenting this one? If yes we need to document the behavior of already mounted csi volumes also.

Yes, I'll add a comment about upgrades to the `requireMsgr2` property so it's more clear. Already mounted csi volumes I would expect to be ok, just like during any other mon failover where the connected clients are updated with the new monmap.

**Comment by travisn** [comment]
@Madhu-1 Regarding the question of whether encryption would be disabled in Rook, it's handled [here](https://github.com/rook/rook/blob/release-1.10/pkg/operator/ceph/cluster/cluster.go#L610-L617) to remove the related ceph settings if encryption is disabled.

**Comment by Madhu-1** [comment]
> @Madhu-1 Regarding the question of whether encryption would be disabled in Rook, it's handled [here](https://github.com/rook/rook/blob/release-1.10/pkg/operator/ceph/cluster/cluster.go#L610-L617) to remove the related ceph settings if encryption is disabled.

Thanks @travisn , even i added that code and totally forgot about it :dagger:

**Pull request merged**